packages/caravan-wallets/src/bitbox.ts

```diff
@@ -0,0 +1,467 @@
+  }
+}
+
+async function convertMultisig(pairedBitBox: PairedBitBox, walletConfig: MultisigWalletConfig): Promise<{ scriptConfig: BtcScriptConfig; keypathAccount: string; }> {
+  const ourRootFingerprint = await pairedBitBox.rootFingerprint();
+
+  const ourXpubIndex = walletConfig.extendedPublicKeys.findIndex(key => key.xfp == ourRootFingerprint);
+  if (ourXpubIndex === -1) {
+    throw new Error('This BitBox02 seems to not be present in the multisig quorum.');
+  }
+  const scriptConfig = {
+    multisig: {
+      threshold: walletConfig.quorum.requiredSigners,
+      scriptType: convertToBtcMultisigScriptType(walletConfig.addressType),
+      xpubs: walletConfig.extendedPublicKeys.map(key => key.xpub),
+      ourXpubIndex,
+    },
+  };
+  const keypathAccount = walletConfig.extendedPublicKeys[ourXpubIndex].bip32Path;
+  return {
+    scriptConfig,
+    keypathAccount,
+  }
+}
+
+/**
+ * Base class for interactions with BitBox hardware wallets.
+ *
+ * Subclasses must implement their own `run()` method.  They may use
+ * the `withDevice` method to connect to the BitBox API.
+ *
+ * @example
+ * import {BitBoxInteraction} from "@caravan/wallets";
+ * // Simple subclass
+ *
+ * class SimpleBitBoxInteraction extends BitBoxInteraction {
+ *
+ *   constructor({param}) {
+ *     super({});
+ *     this.param =  param;
+ *   }
+ *
+ *   async run() {
+ *     return await this.withDevice(async (pairedBitBox) => {
+ *       return pairedBitBox.doSomething(this.param); // Not a real BitBox API call
+ *     });
+ *   }
+ *
+ * }
+ *
+ * // usage
+ * const interaction = new SimpleBitBoxInteraction({param: "foo"});
+ * const result = await interaction.run();
+ * console.log(result); // whatever value `app.doSomething(...)` returns
+ *
+ */
+export class BitBoxInteraction extends DirectKeystoreInteraction {
+  appVersion?: string;
+
+  appName?: string;
+
+  /**
+   * Adds `pending` messages at the `info` level about ensuring the
+   * device is plugged in (`device.connect`) and unlocked
+   * (`device.unlock`).  Adds an `active` message at the `info` level
+   * when communicating with the device (`device.active`).
+   */
+  messages() {
+    const messages = super.messages();
+    messages.push({
+      state: PENDING,
+      level: INFO,
+      text: "Please plug in your BitBox.",
+      code: "device.setup",
+    });
+    messages.push({
+      state: ACTIVE,
+      level: INFO,
+      text: "Communicating with BitBox...",
+      code: "device.active",
+    });
+    return messages;
+  }
+
+  async withDevice<T>(f: (device: PairedBitBox) => Promise<T>): Promise<T> {
+    const bitbox = await import('bitbox-api');
+
+    try {
+      const unpaired = await bitbox.bitbox02ConnectAuto(() => {
+      })
+      const pairing = await unpaired.unlockAndPair()
+      const pairingCode = pairing.getPairingCode()
+      if (pairingCode) {
+        // TODO: display pairing code to the user.
+        console.log('Pairing code:', pairingCode);
```

Comment:
@bucko13 the BitBox02 asks the user to confirm a pairing code to establish the encrypted communication channel. It is known only when running a BitBox02 interaction, at this line. Any suggestions for how to go about showing the code to the user in caravan? Could the messages be used somehow?

Reply:
My first instinct is to use the `messages` api as you suggest. It's honestly not a part of the system I'm super familiar with so I'm not sure if you can run through them asynchronously like that, but I think you can. 

Maybe the way to do it is set up particular message type for pairing. Push the new message on when it's ready. At this point when you have the code you would add it to the messages queue with the appropriate code and type. The consumer of the library would have to know the API and make sure to make use of `messagesFor` correctly and wait for the expected message to be made available.

Reply:
I am not sure how to do that and trigger a re-render of the messages when the pairing code is available and hide it again when it is confirmed. Would it be possible someone on your team who is familiar with this to code the solution in the end?

In the meantime, this PR is still functional and could be merged, just the pairing code would be shown and confirmed on the BitBox02 while the host does not show it. This could be added later by your team in a way you think is proper.

Reply:
If you have the interaction push the messages on the instance array (with the appropriate codes, messages, and tests), that should be sufficient. It can be up to the consumer of the library how they want to track that and update the UI. For a React app, if you track an object via state or properties for example, if the object changes then that should trigger a re-render (or with a useEffect you can trigger some other action to occur). But those are implementation details we can play with later.

Reply:
It seems `messages()` is a function in the interaction, not an instance property/array. How would I trigger clients to call it when the pairing code becomes available? It seems this is not currently supported. 

I could add a "pairingCallback" param, but it would need to be passed to every interaction, since every interaction with the device could potentially ask to pair.

Reply:
Sorry for the delay @benma. 

That's right, it's a function so that the underlying messages can be updated based on different states. You would add the pairing messages dictating a specific state that it should be relevant for. I think it should be fine to build in some UX so that the user clicks a button that prompts user to "confirm pairing code". Clicking that would call the "messagesFor" to get the pairing code. If it's not available, tell the user and re-enable the button. 

Can we at least include in this PR adding the message? I think the rest of the above should also work, but I haven't tried obviously so I might be wrong.

Reply:
I am afraid it won't work this way. The library client (e.g. the coordinator) does not have a way of knowing when the pairing code is being shown, which it needs to know in order to ask the user to confirm. 

I added a commit with a potential solution, please check it out. We can scrap it again if it is not good, it's just a proposal with a working demo:

- All interactions have a `showPairingCode` callback that can be supplied to show/hide the code when the BitBox asks for it
- If the callback is not provided (not implemented), by default a popup is opened to show the code, so at least there a way of checking the code even if the client did not implement a specific way of showing the pairing code.

What do you think?

Reply:
Ok, I think you've convinced me :). Thanks for your patience. Can we clean up the console log now that you've got the callback?